### ISSUE TITLE
Allow pluggable QR code analyzers

### ISSUE THREAD

**Issue by schmich** [schmich/instascan#34]
- https://github.com/LazarSoft/jsqrcode
- https://github.com/zz85/quirc.js
- https://github.com/ghybs/zxingjs2
- Current ZXing
- Create multiple builds?

**Comment by pocesar** [comment]
ZXing is abandoned, maybe use https://github.com/edi9999/jsqrcode?

**Comment by schmich** [comment]
@pocesar Are you seeing any perf/recognition issues with the current setup?

I don't wan't to stick with an abandoned library, but the plan right now is:

- Short-term: stick with the current setup (emscripten'd build of the C++ port of ZXing) unless there are serious issues
- Medium-term: Add adapters for pluggable QR code libraries if there is demand for it
- Long-term: As WebAssembly gets implemented in the various major browsers, investigate integration with the [original Java ZXing implementation](https://github.com/zxing/zxing) via wasm

**Comment by pocesar** [comment]
@schmich actually, I'm not having any issues, the only problem I see, is maintainability. each library have a different API, so you'd need to create a plugin interface between them and release / maintain them.

**Comment by styfle** [comment]
> Long-term: As WebAssembly gets implemented in the various major browsers, investigate integration with the original Java ZXing implementation via wasm

Yes! This would be excellent! I see that today, WASM is available in about [65% of browsers in the wild](https://caniuse.com/#feat=wasm).
If you take into account only the latest version of each browser, then WASM is 100% supported 💯
